src/driver-sinowealth.c

```diff
@@ -47,171 +63,279 @@
+	{ 0x1, 125 },
+	{ 0x2, 250 },
+	{ 0x3, 500 },
+	{ 0x4, 1000 },
+};
+
+/* @return Internal report rate representation or 0 on error. */
+static uint8_t
+sinowealth_report_rate_to_raw(unsigned int report_rate)
 {
-	return (raw + 1) * 100;
+	const struct sinowealth_report_rate_mapping *mapping = NULL;
+	ARRAY_FOR_EACH(sinowealth_report_rate_map, mapping)
+		if (mapping->report_rate == report_rate)
+			return mapping->raw;
+	return 0;
 }
 
-static int
-sinowealth_dpi_to_raw(int dpi)
+/* @return Report rate in hz or 0 on error. */
+static unsigned int
+sinowealth_raw_to_report_rate(uint8_t raw)
 {
-	assert(dpi >= SINOWEALTH_DPI_MIN && dpi <= SINOWEALTH_DPI_MAX);
-	return dpi / 100 - 1;
+	const struct sinowealth_report_rate_mapping *mapping = NULL;
+	ARRAY_FOR_EACH(sinowealth_report_rate_map, mapping)
+		if (mapping->raw == raw)
+			return mapping->report_rate;
+	return 0;
 }
 
-static struct ratbag_color
-sinowealth_raw_to_color(struct sinowealth_rgb8 raw)
+/* @return Maximum DPI for sensor `sensor` or 0 on error. */
+static unsigned int
+get_max_dpi_for_sensor(enum sinowealth_sensor sensor)
+{
+	switch (sensor) {
+	case PWM3327: return 10200;
+	case PWM3360: return 12000;
+	case PWM3389: return 16000;
+	default: return 0;
+	}
+}
+
+/* Convert internal sensor resolution `raw` to DPI. */
+static unsigned int
+sinowealth_raw_to_dpi(struct ratbag_device *device, unsigned int raw)
 {
-	return (struct ratbag_color) {.red = raw.r, .green = raw.g, .blue = raw.b};
+	struct sinowealth_data *drv_data = device->drv_data;
+	enum sinowealth_sensor sensor = drv_data->sensor;
+
+	if (sensor == PWM3360)
+		raw += 1;
+
+	unsigned int dpi = raw * 100;
+
+	return dpi;
 }
 
-static struct sinowealth_rgb8
-sinowealth_color_to_raw(struct ratbag_color color)
+/* Convert DPI `dpi` to internal sensor resolution. */
+static unsigned int
+sinowealth_dpi_to_raw(struct ratbag_device *device, unsigned int dpi)
 {
-	return (struct sinowealth_rgb8) {.r = color.red, .g = color.green, .b = color.blue};
+	struct sinowealth_data *drv_data = device->drv_data;
+	enum sinowealth_sensor sensor = drv_data->sensor;
+
+	assert(dpi >= SINOWEALTH_DPI_MIN && dpi <= get_max_dpi_for_sensor(sensor));
+
+	unsigned int raw = dpi / 100;
+
+	if (sensor == PWM3360)
+		raw -= 1;
+
+	return raw;
 }
 
+/* Convert internal mouse color `raw` to color.
+ * If LED type defined in the device data is incorrect, RBG color order is used.
+ */
 static struct ratbag_color
-sinowealth_rbg_to_color(struct sinowealth_rbg8 raw)
+sinowealth_raw_to_color(struct ratbag_device *device, struct sinowealth_color raw)
 {
-	return (struct ratbag_color) {.red = raw.r, .green = raw.g, .blue = raw.b};
+	struct sinowealth_data *drv_data = device->drv_data;
+
+	switch (drv_data->led_type) {
+	case LED_RGB: return (struct ratbag_color){ .red = raw.data[0], .green = raw.data[1], .blue = raw.data[2] };
+	default:
+	case LED_RBG: return (struct ratbag_color){ .red = raw.data[0], .green = raw.data[2], .blue = raw.data[1] };
+	}
```

Comment:
I think this would be a lot more readable with a temporary variable for the `ratbag_color` to return.

Reply:
Fixed. Thank you!